src/NServiceBus.Core/Pipeline/PipelineExecutionExtensions.cs

```diff
@@ -64,29 +67,41 @@ public static Delegate CreatePipelineExecutionExpression(this IBehavior[] behavi
         }
 
         /// <code>
-        /// context{i} => ((BehaviorType)context{i}.Extensions.Behaviors[i]).Invoke(context{i+1} => previous)
+        /// context{i} => GetBehavior(context{i}, {i}).Invoke(context{i+1} => previous)
         /// </code>>
         static Delegate CreateBehaviorCallDelegate(MethodInfo methodInfo, ParameterExpression outerContextParam, Type behaviorType, Delegate previous, int i, List<Expression> expressions = null)
         {
-            PropertyInfo extensionProperty = typeof(IExtendable).GetProperty("Extensions");
-            Expression extensionPropertyExpression = Expression.Property(outerContextParam, extensionProperty);
-            PropertyInfo behaviorsProperty = typeof(ContextBag).GetProperty("Behaviors", BindingFlags.Instance | BindingFlags.NonPublic);
-            Expression behaviorsPropertyExpression = Expression.Property(extensionPropertyExpression, behaviorsProperty);
-            Expression indexerPropertyExpression = Expression.ArrayIndex(behaviorsPropertyExpression, Expression.Constant(i));
-            Expression castToBehavior = Expression.Convert(indexerPropertyExpression, behaviorType);
-            Expression body = Expression.Call(castToBehavior, methodInfo, outerContextParam, Expression.Constant(previous));
+            MethodInfo getBehaviorMethodInfo = typeof(PipelineExecutionExtensions).GetMethod("GetBehavior")!.MakeGenericMethod(outerContextParam.Type, behaviorType);
+            Expression getBehaviorCallExpression = Expression.Call(null, getBehaviorMethodInfo, outerContextParam, Expression.Constant(i));
+            Expression body = Expression.Call(getBehaviorCallExpression, methodInfo, outerContextParam, Expression.Constant(previous));
             var lambdaExpression = Expression.Lambda(body, outerContextParam);
             expressions?.Add(lambdaExpression);
             return lambdaExpression.CompileFast();
         }
 
+#if NET
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Does aggressive inlining have any effect since the method is only found via reflection? Or does this affect the compilation of the lambda expression as well?

Reply:
I was not able to find any documentation whether expression trees that point to static methods that are aggressively inlined can be properly JITed. I did run tests with and without the inlining hint and I saw 3-5% gains when I added the inlining hint. Unfortunately, I do not have the test results on the laptop I'm currently typing on, but I'm happy to share it later in case you are interested.